File: 2015/j3/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from pathlib import Path

def read_input(file_name: str):
    path = Path(__file__).resolve().parent / file_name
    with path.open("r", encoding="utf-8") as f:
        return [line.rstrip("\n") for line in f]

def part1():
    with open("input.txt") as f:
        moves = f.read().strip()

    x, y = 0, 0
    visited = {(x, y)}

    for move in moves:
        if move == "^":
            y += 1
        elif move == "v":
            y -= 1
        elif move == ">":
            x += 1
        elif move == "<":
            x -= 1
        visited.add((x, y))

    return len(visited)

def part2():
    with open("input.txt") as f:
        moves = f.read().strip()

    santa_x, santa_y = 0, 0
    robo_x, robo_y = 0, 0

    visited = {(0, 0)}

    for idx, move in enumerate(moves):
        if idx % 2 == 0:
            if move == "^":
                santa_y += 1
            elif move == "v":
                santa_y -= 1
            elif move == ">":
                santa_x += 1
            elif move == "<":
                santa_x -= 1
            visited.add((santa_x, santa_y))
        else:
            if move == "^":
                robo_y += 1
            elif move == "v":
                robo_y -= 1
            elif move == ">":
                robo_x += 1
            elif move == "<":
                robo_x -= 1
            visited.add((robo_x, robo_y))

    return len(visited)

def main():
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

if __name__ == "__main__":
    main()

File: 2015/j3/main.kt
package aoc

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

fun readInput(fileName: String): List<String> {
    val path: Path = Paths.get("").resolve(fileName)
    return Files.readAllLines(path)
}

fun part1(): Any {
    val moves = readInput("input.txt")[0]

    var x = 0
    var y = 0
    val visited = mutableSetOf(Pair(x, y))

    for (move in moves) {
        when (move) {
            '^' -> y += 1
            'v' -> y -= 1
            '>' -> x += 1
            '<' -> x -= 1
        }
        visited.add(Pair(x, y))
    }

    return visited.size
}

fun part2(): Any {
    val moves = readInput("input.txt")[0]

    var santaX = 0
    var santaY = 0
    var roboX = 0
    var roboY = 0

    val visited = mutableSetOf(Pair(0, 0))

    for ((index, move) in moves.withIndex()) {
        if (index % 2 == 0) {
            when (move) {
                '^' -> santaY += 1
                'v' -> santaY -= 1
                '>' -> santaX += 1
                '<' -> santaX -= 1
            }
            visited.add(Pair(santaX, santaY))
        } else {
            when (move) {
                '^' -> roboY += 1
                'v' -> roboY -= 1
                '>' -> roboX += 1
                '<' -> roboX -= 1
            }
            visited.add(Pair(roboX, roboY))
        }
    }

    return visited.size
}

fun main() {
    println("Part 1: ${part1()}")
    println("Part 2: ${part2()}")
}
